plugins/deisa/deisa.cxx

```diff
@@ -0,0 +1,212 @@
+	long m_rank;
+	long m_size;
+	Expression m_time_step;
+	py::object m_bridge;
+
+public:
+	static std::pair<std::unordered_set<std::string>, std::unordered_set<std::string>> dependencies() { return {{"mpi"}, {"mpi"}}; }
+
+	deisa_plugin(Context& ctx, PC_tree_t conf)
+		: Plugin{ctx}
+	{
+		if (!Py_IsInitialized()) {
+			py::initialize_interpreter();
+			m_interpreter_initialized_in_plugin = true;
+		}
+
+		// init params
+		each(conf, [&](PC_tree_t key_tree, PC_tree_t value) {
+			std::string key = to_string(key_tree);
+			if (key == "scheduler_info") {
+				m_scheduler_info = to_string(value);
+			} else if (key == "deisa_arrays") {
+				each(value, [&](PC_tree_t key_map, PC_tree_t value_map) { m_deisa_arrays.emplace(to_string(key_map), ctx.datatype(value_map)); });
+			} else if (key == "time_step") {
+				m_time_step = to_string(value);
+			} else if (key == "map_in") {
+				//
+			} else if (key == "logging" || key == "init_on") {
+				//
+			} else {
+				throw Config_error{key_tree, "Unknown key in Deisa file configuration: '{}'", key};
+			}
+		});
+
+
+		int mpi_size;
+		m_rank = Ref_r{ctx.desc("MPI_COMM_WORLD_rank").ref()}.scalar_value<long>();
+		MPI_Comm comm = *static_cast<const MPI_Comm*>(Ref_r{ctx.desc("MPI_COMM_WORLD").ref()}.get());
+		MPI_Comm_size(comm, &mpi_size);
+		m_size = static_cast<long>(mpi_size);
+
+
+		// plugin init
+		PC_tree_t init_tree = PC_get(conf, ".init_on");
+		if (!PC_status(init_tree)) {
+			ctx.callbacks().add_event_callback([&](const std::string&) { init_deisa(); }, to_string(init_tree));
+		} else {
+			throw Config_error{conf, "Deisa plugin requires init_on key "};
+		} // TODO: replace with try/catch when #480 is fixed.
+
+		// map_in
+		PC_tree_t map_in = PC_get(conf, ".map_in");
+		if (!PC_status(map_in)) {
+			each(map_in, [&](PC_tree_t key_map, PC_tree_t value_map) { // TODO: when #481 is fixed, use `or_none` variant.
+				ctx.callbacks().add_data_callback(
+					[&, deisa_array_name = to_string(value_map)](const std::string&, const Ref& data_ref) {
+						try {
+							assert(hasattr(m_bridge obj, "publish_data"));
+							py::object publish_data = m_bridge.attr("publish_data");
+#ifdef NDEBUG
+							publish_data(to_python(data_ref), deisa_array_name.c_str(), m_time_step.to_long(ctx), "debug"_a = false);
+#else
+							publish_data(to_python(data_ref), deisa_array_name.c_str(), m_time_step.to_long(ctx), "debug"_a=true);
+#endif
+						} catch (const std::exception& e) {
+							throw Plugin_error("While publishing data. Caught exception: {}", std::string(e.what()));
+						} catch (...) {
+							throw Plugin_error("While publishing data.");
+						}
+					},
+					to_string(key_map)
+				);
+			});
+		}
+	}
+
+	~deisa_plugin() noexcept override
+	{
+		try {
+			if (m_interpreter_initialized_in_plugin) {
+				if (hasattr(m_bridge, "release")) {
+					// call bridge release() so that we can clear things before destructor is called (if it is ever called !).
+					m_bridge.release();
```

Comment:
please set it to null afterward, so that the garbage collector has a chance to run.

Also, this can be called whether we created the interpreter or not, no?

Reply:
yes `release` can be called anytime if the interpreter is still alive.